### PR TITLE
Feature/flash donor

### PR DESCRIPTION
**Pull request by Rup-Narayan-Rajbanshi**
Addresses Flash Update

## Changes

Share flash update to groups of Donors
    - Add table FlashUpdateShare
    - Add api to share flashupdate to donors
    - Add flashUpdateSerializer
    - Add test case for flash update share

## Checklist 
Things that should succeed before merging.

- [x] Updated/ran unit tests
- [ ] Updated CHANGELOG.md

## Release

If there is a version update, make sure to tag the repository with the latest version.

### CODE REVIEW

File: flash_update/models.py [hunk 1]
```diff
@@ -208,12 +208,30 @@ def __str__(self):
...
+    groups = models.ManyToManyField(DonorGroup, verbose_name=_('donor group'), blank=True)
 
     def __str__(self):
         return self.organization_name
+
+
+class FlashUpdateShare(models.Model):
+    flash_update = models.ForeignKey(FlashUpdate, on_delete=models.CASCADE, related_name='flash_update_share')
+    donors = models.ManyToManyField(Donors, blank=True, null=True)
+    donor_groups = models.ManyToManyField(DonorGroup, blank=True, null=True)
```
Comment (thenav56) [inline flash_update/models.py:233]: null=True not required.

File: flash_update/utils.py [hunk 2]
```diff
@@ -13,14 +13,16 @@
...
+    try:
+        pdf = pisa.pisaDocument(BytesIO(html.encode("ISO-8859-1")), result)
+        if not pdf.err:
+            file = {
+                'filename': 'flash_update.pdf',
+                'file': result.getvalue()
+            }
+            return file
+    except:
+        return None
```
Comment (thenav56) [inline flash_update/utils.py:25]: Log the error here logging.error('message', exc_info=True).

**Review by thenav56 — CHANGES_REQUESTED** [review]

File: flash_update/utils.py [hunk 3]
```diff
@@ -84,3 +91,19 @@ def send_email_when_flash_update_created(instance):
...
+    donor_groups_emails = list(Donors.objects.filter(groups__id__in=doner_groups).values_list('email', flat=True))
+    users_emails = donors_emails + donor_groups_emails
+    send_notification(
+        'Flash Update',
+        users_emails,
+        render_to_string('email/flash_update/donor_email.html', email_context),
+        'Flash Update',
+        [render_to_pdf('email/flash_update/flash_pdf.html', email_context)]
+    )
+    return email_context
```
Comment (thenav56) [inline flash_update/utils.py:109]:
```suggestion
def share_flash_update(instance):
    email_context = get_email_context(instance.flash_update)
    donors_emails = instance.donors.values_list('email', flat=True)
    donor_groups_emails = Donors.objects\
        .filter(groups__in=instance.donor_groups.all())\
        .values_list('email', flat=True)
    users_emails = set([*donors_emails, *donor_groups_emails])
    send_notification(
        'Flash Update',
        users_emails,
        render_to_string('email/flash_update/donor_email.html', email_context),
        'Flash Update',
        [render_to_pdf('email/flash_update/flash_pdf.html', email_context)]
    )
    return email_context

```

**Review by thenav56 — CHANGES_REQUESTED** [review]

File: flash_update/serializers.py [hunk 4]
```diff
@@ -18,7 +18,10 @@
     FlashCountryDistrict,
     FlashGraphicMap,
     FlashAction,
-    FlashActionsTaken
+    FlashActionsTaken,
+    DonorGroup,
+    Donors,
+    FlashUpdateShare
```
Comment (thenav56) [inline flash_update/serializers.py:24]: Let's always add `,` for the last item in the list as well. This creates a better diff for future changes.

File: flash_update/test_views.py [hunk 5]
```diff
@@ -5,13 +5,15 @@
...
 import api.models as models
-from flash_update.models import FlashUpdate, FlashEmailSubscriptions, FlashGraphicMap
+from flash_update.models import FlashUpdate, FlashEmailSubscriptions, FlashGraphicMap, FlashUpdateShare
 from flash_update.factories.flash_update import (
     FlashUpdateFactory,
     FlashGraphicMapFactory,
-    FlashActionFactory
+    FlashActionFactory,
+    DonorFactory,
+    DonorGroupFactory
```
Comment (thenav56) [inline flash_update/test_views.py:14]: Let's always add `,` for the last item in the list as well. This creates a better diff for future changes.

File: flash_update/views.py [hunk 6]
```diff
@@ -14,11 +14,17 @@
 from .models import (
     FlashUpdate,
     FlashGraphicMap,
-    FlashAction
+    FlashAction,
+    DonorGroup,
+    Donors,
+    FlashUpdateShare
```
Comment (thenav56) [inline flash_update/views.py:20]: Let's always add `,` for the last item in the list as well. This creates a better diff for future changes.

File: flash_update/views.py [hunk 7]
```diff
@@ -14,11 +14,17 @@
...
 )
 from .serializers import (
     FlashUpdateSerializer,
-    FlashGraphicMapSerializer
+    FlashGraphicMapSerializer,
+    DonorGroupSerializer,
+    DonorsSerializer,
+    ShareFlashUpdateSerializer
```
Comment (thenav56) [inline flash_update/views.py:27]: Let's always add `,` for the last item in the list as well. This creates a better diff for future changes.

File: flash_update/admin.py [hunk 8]
```diff
@@ -10,7 +10,9 @@
     FlashActionsTaken,
     FlashCountryDistrict,
     FlashEmailSubscriptions,
-    Donors
+    Donors,
+    DonorGroup,
+    FlashUpdateShare
```
Comment (thenav56) [inline flash_update/admin.py:15]: Let's always add `,` for the last item in the list as well. This creates a better diff for future changes.

File: flash_update/factories/flash_update.py [hunk 9]
```diff
@@ -10,7 +10,9 @@
 from flash_update.models import (
     FlashUpdate,
     FlashGraphicMap,
-    FlashAction
+    FlashAction,
+    Donors,
+    DonorGroup
```
Comment (thenav56) [inline flash_update/factories/flash_update.py:15]: Let's always add `,` for the last item in the list as well. This creates a better diff for future changes.

**Review by thenav56 — CHANGES_REQUESTED** [review]
Let's always add `,` for the last item in the list as well. This creates a better diff for future changes.

**Review by thenav56 — APPROVED** [review]

**Review by szabozoltan69 — CHANGES_REQUESTED** [review]
Somehow on the admin interface we can see "Donorss" instead of "Donors".
Could you please fix it?
kép

**Comment by szabozoltan69** [comment]
I suggest this to models.py (class Donors), L225:
```
    class Meta:
        verbose_name = _('donor')
```

**Comment by szabozoltan69** [comment]
Shall I insert the above two rows or do you, @Rup-Narayan-Rajbanshi ?

**Review by szabozoltan69 — APPROVED** [review]

**Comment by szabozoltan69** [comment]
Nice tests, @Rup-Narayan-Rajbanshi !
(I fixed the Donorss.)
Deployed to staging.

**Comment by Rup-Narayan-Rajbanshi** [comment]
> Shall I insert the above two rows or do you, @Rup-Narayan-Rajbanshi ?

I will do it @szabozoltan69 .

**Comment by Rup-Narayan-Rajbanshi** [comment]
> Shall I insert the above two rows or do you, @Rup-Narayan-Rajbanshi ?

Sorry for being late. I was busy in some other task. Thanks for your effort

**Pull request merged**